crates/api/src/runtime.rs

```diff
@@ -309,6 +312,12 @@ impl Config {
         self.cache_config = wasmtime_environ::CacheConfig::from_file(None)?;
         Ok(self)
     }
+
+    /// Sets a custom memory allocator
+    pub fn with_host_memory(&mut self, allocator: Arc<dyn Allocator>) -> &mut Self {
```

Comment:
FWIW this is making the `wasmtime_runtime` crate a public dependency of the `wasmtime` crate. We've previously avoided doing this because the `wasmtime_*` family of crate are intended to be far more lower level than the `wasmtime` crate.

That being said we haven't quite had a vector of customization like this before where completely dynamic behavior is needed. I suspect we're going to have these grow over time, but I figured it's worth pointing out that this has a few consequences for the `wasmtime` crate:

* This primarily means that if `wasmtime_runtime` has a major version bump, we have to bump the major version of `wasmtime` as well. It's expected that `wasmtime_runtime` will have *lots* of major version bumps because it's just a bunch of internals. For `wasmtime`, however, I think the idea is that eventually we reach a steady state where we just release new patch/minor versions.

* This is also relatively difficult to use since you'll have to match up, manually, with the version of `wasmtime_runtime` that `wasmtime` is using. 

All that is to say that if we can avoid exposing this degree of a public dependency I think we should. One way, for example, would be to define traits in `wasmtime` as well as in `wasmtime_runtime`. The `wasmtime` crate would then provide its own implementation of the traits in `wasmtime_runtime` for provided types. This is a bit wonky, however, from our perspective and may not be the most trivial to manage.

Reply:
Thank you for the info. I was meaning to ask about it as well.
I'll try to move things around, but first I'd rather come to a consensus on the general picture of how the whole thing should work internally.

There is also a question of MemoryPlan which resides in environ now.